CHEF-4086: Use "Host:" header value to construct returned URLs

hosh opened the pull request:
- CHEF-4086: Use "Host:" header value to construct returned URLs

Requires:
- https://github.com/opscode/chef-pedant/pull/59
- https://github.com/opscode/chef_objects/pull/50
- https://github.com/opscode/omnibus-chef-server/pull/19

Notes:
- I did not include the case where Hosts: is empty (and not missing).
- How to even test this? Pedant won't catch all the cases, so additional unit testing is required.

hosh committed:
CHEF-4086: Use "Host:" header value to construct returned URLs

Erchef needs to send back URLs for resources such as cookbook versions,
clients, etc. However, if it sits behind a firewall or a proxy, then it
might not have access to the name the client used to access the server.

This patch creates the base url by using the value set by the client in
the "Hosts:" header. If the value is missing, then it will fallback on
using the webmachine-provided host name.

There are some security concerns with this approach. One way to mitigate
this is by checking the value against a configurable whitelist.

hosh commented:
WIP. Ping: @seth @sdelano @manderson26 @marcparadise

seth commented:
Code change looks good. I think you could test a fair amount of this with a eunit test in this repo and some use of meck to mock out wrq module.

seth commented on src/chef_wm_util.erl:
```diff
@@ -52,10 +52,17 @@ base_mods() ->
+%% @doc Returns the host from the request headers, or use the webmachine default
+vhost_or_peer(Req) ->
+    case wrq:get_req_header("host", Req) of
+        B when is_binary(B) -> binary_to_list(B);
```
Can get_req_header ever return binary?

hosh replied:
I don't know. I pulled it from the GetHeader fun.

seth commented on src/chef_wm_util.erl:
```diff
@@ -52,10 +52,17 @@ base_mods() ->
+    case wrq:get_req_header("host", Req) of
+        B when is_binary(B) -> binary_to_list(B);
+        S when is_list(S) -> S;
+        undefined ->
```
Perhaps we should capture the empty case here along these lines:

```
None when None == undefined orelse None == "" orelse None == <<"">> ->
```

and have that be the first clause

seth replied:
Although for HTTP 1.1 a missing Host header is cause for 400.

hosh replied:
Huh, I didn't know that Host header is required for HTTP 1.1. I was going to capture the empty case too, but if HTTP 1.1 requires it, is it worth adding this kind of defensive coding?

hosh committed:
Configurable base_resource_url and pass to S3 presigned url generation

 - Make vhost or fqdn url generation configurable
 - Pass on vhost/fqdn when generating S3 presigned url for sandbox,
   depsolver, and cookbook_version

If base_resource_url is set to the atom, 'host_header' (default) then
the URL will use the Host: header value.

Otherwise, it will use the url set for base_resource_url

Further, pass this value to the S3 presigned URL generation.

seth commented on src/chef_wm_util.erl:
```diff
@@ -52,10 +52,29 @@ base_mods() ->
-    Scheme ++ "://" ++ Host ++ PortString.
+    scheme(Req) ++ "://" ++ vhost_or_fqdn(Req).
+
+%% @doc Returns the vhost or fqdn based upon configuration
```
I think here we should either use the Host header or use an explicit URL from config. Probably could simply remove the fallback to wrq:host_tokens - really not sure that is doing anything other than reading from the host header anyhow (could test that).

hosh replied:
I thought the original problem was that wrq:host_tokens ended up getting the info from get_peer() and not at all from the Host: ?

seth commented on src/chef_wm_util.erl:
```diff
@@ -52,10 +53,31 @@ base_mods() ->
+    case application:get_env(chef_wm, base_resource_url) of
+        undefined         -> vhost(Req);
+        {ok, host_header} -> vhost(Req);
+        {ok, ExternalUrl} -> ExternalUrl
```
Maybe `ExternalHost` would be a better name? The URL is constructed in `base_uri`.

hosh replied:
Oh, it would be the host. Ok, yeah, ExternalHost would be a better name. I will need to revise the various commits and comments about that since it says to configure with "http://api.example.com" rather than "api.example.com"

hosh replied:
Though now that I think about it, I think keeping s3_external_url, s3_url, and base_resource_url all URLs instead of host would help keep things consistent. So I'll merge vhost_or_fqdn() with base_uri. It isn't as if we're calling vhost_or_fqdn elsewhere.

seth commented on src/chef_wm_util.erl:
```diff
@@ -52,10 +53,31 @@ base_mods() ->
+%% @doc Returns the host from the request headers, or use the webmachine default
+vhost(Req) ->
+    case wrq:get_req_header("host", Req) of
+        B when is_binary(B) -> binary_to_list(B);
```
I'm nearly certain that wrq:get_req_header returns `string() | undefined`, but I think it is fine to leave this.

seth commented:
This is looking good. I like the organization into small helper functions, makes the intent clear. See the small comment about naming.

hosh committed:
Added tests for chef_wm_util:base_uri()

hosh commented:
Ping: @seth @oferrigni @manderson26 @marcparadise @sdelano
